cranelift/codegen/src/data_value.rs

```diff
@@ -152,15 +192,77 @@ impl DataValue {
+            types::I64 => DataValue::I64(i64::from_le_bytes(src[..8].try_into().unwrap())),
+            types::I128 => DataValue::I128(i128::from_le_bytes(src[..16].try_into().unwrap())),
+            types::F32 => DataValue::F32(Ieee32::with_bits(u32::from_le_bytes(
+                src[..4].try_into().unwrap(),
+            ))),
+            types::F64 => DataValue::F64(Ieee64::with_bits(u64::from_le_bytes(
+                src[..8].try_into().unwrap(),
+            ))),
+            _ if ty.is_vector() => {
+                if ty.bytes() == 16 {
+                    DataValue::V128(src[..16].try_into().unwrap())
+                } else if ty.bytes() == 8 {
+                    DataValue::V64(src[..8].try_into().unwrap())
+                } else {
+                    unimplemented!()
+                }
+            }
+            _ => unimplemented!(),
+        }
+    }
+
+    /// Read a [DataValue] from a slice using a given [Type] in big-endian byte order.
+    ///
+    /// # Panics:
+    ///
+    /// Panics if the slice does not have enough space to accommodate the [DataValue]
+    pub fn read_from_slice_be(src: &[u8], ty: Type) -> Self {
+        match ty {
+            types::I8 => DataValue::I8(i8::from_be_bytes(src[..1].try_into().unwrap())),
+            types::I16 => DataValue::I16(i16::from_be_bytes(src[..2].try_into().unwrap())),
+            types::I32 => DataValue::I32(i32::from_be_bytes(src[..4].try_into().unwrap())),
+            types::I64 => DataValue::I64(i64::from_be_bytes(src[..8].try_into().unwrap())),
+            types::I128 => DataValue::I128(i128::from_be_bytes(src[..16].try_into().unwrap())),
+            types::F32 => DataValue::F32(Ieee32::with_bits(u32::from_be_bytes(
+                src[..4].try_into().unwrap(),
+            ))),
+            types::F64 => DataValue::F64(Ieee64::with_bits(u64::from_be_bytes(
+                src[..8].try_into().unwrap(),
+            ))),
+            _ if ty.is_vector() => {
+                if ty.bytes() == 16 {
+                    DataValue::V128(src[..16].try_into().unwrap())
+                } else if ty.bytes() == 8 {
+                    DataValue::V64(src[..8].try_into().unwrap())
+                } else {
+                    unimplemented!()
+                }
+            }
+            _ => unimplemented!(),
+        }
+    }
+
+    /// Write a [DataValue] to a memory location in native-endian byte order.
     pub unsafe fn write_value_to(&self, p: *mut u128) {
         let size = self.ty().bytes() as usize;
-        self.write_to_slice(std::slice::from_raw_parts_mut(p as *mut u8, size));
+        self.write_to_slice_ne(std::slice::from_raw_parts_mut(p as *mut u8, size));
     }
 
-    /// Read a [DataValue] from a memory location using a given [Type].
+    /// Read a [DataValue] from a memory location using a given [Type] in native-endian byte order.
     pub unsafe fn read_value_from(p: *const u128, ty: Type) -> Self {
```

Comment:
As far as I can see, this function is only ever used for testing purposes so I assumed that specific endianness is not an issue here, but do let me know.

Reply:
That seems to be used by our trampolines, so _ne should be the correct choice!